contracts/infrastructure/DappRegistry.sol

```diff
@@ -0,0 +1,78 @@
+pragma solidity ^0.6.12;
+
+import "./IAuthoriser.sol";
+import "./dapp/IFilter.sol";
+import "./base/Owned.sol";
+import "./storage/Storage.sol";
+
+contract DappRegistry is IAuthoriser, Storage, Owned {
+
+    bytes32 constant internal ARGENT_DEFAULT = keccak256("argent_default");
+
+    struct Authorisation {
+        bool isActive;
+        address filter;
+    }
+    
+    // mapping [wallet][name][status]
+    mapping (address => mapping (bytes32 => bool)) public isEnabled;
+    // mapping [name][address][authorisation]
+    mapping (bytes32 => mapping (address => Authorisation)) public authorisations;
+
+    bytes32[] public registries;
+
+    mapping (bytes32 => address) public registryManager;
+
+    event RegistryCreated(bytes32 registry, address manager);
+    event RegistryRemoved(bytes32 registry);
+
+    function authorise(address _wallet, address _contract, bytes calldata _data) external view override returns (bool) {
+        (bool isActive, address filter) = getFilter(_wallet, _contract);
+        if (isActive) {
+            return _data.length == 0 || filter == address(0) || IFilter(filter).validate(_data);
+        }
+        return false;
+    }
+
+    function toggle(address _wallet, bytes32 _registry) external override onlyModule(_wallet) returns (bool) {
+        require(_registry == ARGENT_DEFAULT || registryManager[_registry] != address(0), "AR: unknow registry");
+        bool current = isEnabled[_wallet][_registry];
+        isEnabled[_wallet][_registry] = !current;
+        return !current;
```

Comment:
This should probably be `return (_registry == ARGENT_DEFAULT) ? current : !current` to make the `ToggledDappRegistry(_wallet, _registry, isEnabled)` event easier to interpret for clients.